ouroboros-network-protocols/src/Ouroboros/Network/Protocol/ObjectDiffusion/Type.hs

```diff
@@ -0,0 +1,359 @@
+  -- identifiers that it wishes to submit.
+  --
+  -- There are two sub-states for this, for blocking and non-blocking cases.
+  StObjectIds :: StBlockingStyle -> ObjectDiffusion initAgency objectId object
+  -- | The outbound node has agency; it must reply with the list of
+  -- objects.
+  StObjects :: ObjectDiffusion initAgency objectId object
+  -- | Nobody has agency; termination state.
+  StDone :: ObjectDiffusion initAgency objectId object
+
+instance
+  ( ShowProxy objectId,
+    ShowProxy object
+  ) =>
+  ShowProxy (ObjectDiffusion objectId object)
+  where
+  showProxy _ =
+    concat
+      [ "ObjectDiffusion ",
+        showProxy (Proxy :: Proxy objectId),
+        " ",
+        showProxy (Proxy :: Proxy object)
+      ]
+
+instance ShowProxy (StIdle :: ObjectDiffusion initAgency objectId object) where
+  showProxy _ = "StIdle"
+
+type SingObjectDiffusion ::
+  ObjectDiffusion initAgency objectId object ->
+  Type
+data SingObjectDiffusion k where
+  SingInit :: SingObjectDiffusion StInit
+  SingIdle :: SingObjectDiffusion StIdle
+  SingObjectIds ::
+    SingBlockingStyle stBlocking ->
+    SingObjectDiffusion (StObjectIds stBlocking)
+  SingObjects :: SingObjectDiffusion StObjects
+  SingDone :: SingObjectDiffusion StDone
+
+deriving instance Show (SingObjectDiffusion k)
+
+instance StateTokenI StInit where stateToken = SingInit
+
+instance StateTokenI StIdle where stateToken = SingIdle
+
+instance
+  (SingI stBlocking) =>
+  StateTokenI (StObjectIds stBlocking)
+  where
+  stateToken = SingObjectIds sing
+
+instance StateTokenI StObjects where stateToken = SingObjects
+
+instance StateTokenI StDone where stateToken = SingDone
+
+data StBlockingStyle where
+  -- | In this sub-state the reply need not be prompt. There is no timeout.
+  StBlocking :: StBlockingStyle
+  -- | In this state the peer must reply. There is a timeout.
+  StNonBlocking :: StBlockingStyle
+
+newtype NumObjectIdsToAck = NumObjectIdsToAck {getNumObjectIdsToAck :: Word16}
+  deriving (Eq, Ord, NFData, Generic)
+  deriving newtype (Num, Enum, Real, Integral, Bounded, NoThunks)
+  deriving (Semigroup) via (Sum Word16)
+  deriving (Monoid) via (Sum Word16)
+  deriving (Show) via (Quiet NumObjectIdsToAck)
+
+newtype NumObjectIdsToReq = NumObjectIdsToReq {getNumObjectIdsToReq :: Word16}
+  deriving (Eq, Ord, NFData, Generic)
+  deriving newtype (Num, Enum, Real, Integral, Bounded, NoThunks)
+  deriving (Semigroup) via (Sum Word16)
+  deriving (Monoid) via (Sum Word16)
+  deriving (Show) via (Quiet NumObjectIdsToReq)
+
+-- | There are some constraints of the protocol that are not captured in the
+-- types of the messages, but are documented with the messages. Violation
+-- of these constraints is also a protocol error. The constraints are intended
+-- to ensure that implementations are able to work in bounded space.
+instance Protocol (ObjectDiffusion initAgency objectId object) where
+  -- \| The messages in the object diffusion protocol.
```

Comment:
```suggestion
  -- | The messages in the object diffusion protocol.
```
I know haddock doesn't support it yet, but it also doesn't complain about it, doesn't it?